### DISCUSSION TITLE
Time Steps and Interpolation

### DISCUSSION THREAD

**Discussion by a-day-old-bagel** [jrouwe/JoltPhysics#502] [Q&A] [answered]
I didn't see an existing question about this, although I probably missed it since this seems kind of important (?) and FAQ-ish.

Anyway, let's say I want to update the physics at a fixed step, but I want to render more frequently than that. The usual approach is to interpolate between physics steps to produce the rendering transforms. When I used Bullet, that interpolation was built-in for you, so that one of the getters for a body's transform returned the interpolated value iirc.

Is this included in Jolt in any way, or am I going to have to do all of the interpolation myself outside of Jolt?

**Comment by jrouwe** [comment, marked as answer]
Interpolation is not part of Jolt and needs to be implemented in your 'game' code. It's not hard to do this, let me know if I need to explain how.

**Reply by a-day-old-bagel** [reply]
I'm assuming it's just the usual, where you have two physics frames, resulting from two adjacent simulation steps, and you have the body transforms stored from each of them. Then for each body you want to render, you interpolate the position vector linearly between the two given the renderer's current "in-between" time, and interpolate the quaternions similarly?

If you have suggestions as to how best to store the two end points of the interpolation (or fetch them instead of storing) or any other insights, (or any other correction or clarification you'd like to make) I'd be happy to hear them.

EDIT:
I guess an addendum: Would it be inadvisable to try the following:
Instead of storing the two "endpoints" of the interpolation as a bunch of vec3's and quats somewhere in the game code, would it be an ok idea to have a separate physics system running on your graphics thread, where you `restoreState` the initial interpolation endpoint into it, run it for a small partial time step to accomplish the interpolation (with added benefit of collisions), and pull the results directly from that partial step to render? Does that make any sense? I suppose this would qualify as extrapolation instead of interpolation, though.

EDIT EDIT:
Or even have the graphics thread sub-stepping physics system turn off all collisions so that it's cheaper, and just use it for approximate (potentially non-correct) interpolation, with the only added benefit I can think of now being the ability to see non-linear interpolations resulting from active accelerations or something. Actually, I guess you'd want to turn off gravity too, so that bodies resting on a floor don't keep trying to fall through it in the interpolation. Maybe there are more problems. Maybe storing the two endpoints' vecs and quats separately is safer, idk.

**Comment by jrouwe** [comment]
Assuming that your 'game object' has at most 1 Body associated, I think the simplest way would be that your 'game object' contains Vec3 mPreviousPosition and Quat mPreviousRotation rotation. If you move a Body yourself by setting it's position/rotation you need to update mPreviousPosition/Rotation too. The Body should point to your 'game object' using its mUserData. At the beginning of each step you would iterate the list returned by PhysicsSystem::GetActiveBodies and store the current Bodies position/rotation in mPreviousPosition/Rotation. You need to register a BodyActivationListener and listen to any body that gets a OnBodyDeactivated callback and set mPreviousPosition/Rotation to the current Body position/rotation at the beginning of the next time step too (or you can do it straight away, the movement should be small so you probably won't notice that it doesn't interpolate for 1 frame). Result is that at the end of the time step all your game objects store the correct previous position/rotation and the Body contain the new position/rotation. When you render you can interpolate between these 2 values.

> would it be an ok idea to have a separate physics system running on your graphics thread

This is a possibility, but waaaay more expensive.

> Or even have the graphics thread sub-stepping physics system turn off all collisions so that it's cheaper

This sounds like it is going to create a lot of artifacts. If you don't want to store state you can also just calculate ```current_position - velocity * delta_time``` to extrapolate backwards with the same kind of artifacts.

**Reply by a-day-old-bagel** [reply]
Thank you for your time, this has been helpful.
